### PR TITLE
Support SVG element

### LINKED ISSUE #12974: Support SVG element
Spec: https://www.w3.org/TR/SVG11/struct.html#SVGElement

Goals:
* create a `SVGElement` type that is created by the parser
* support the `width` and `height` attributes
* use these attributes to create a new fragment in layout with the appropriate width and height

Blocks #12973.

### PR DESCRIPTION
**Pull request by splav**
## minimal SVG element implementation

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #12974 (github issue number if applicable).

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13589)

### CODE REVIEW

File: components/layout/fragment.rs [hunk 1]
```diff
@@ -155,6 +156,7 @@ pub enum SpecificFragmentInfo {
     Iframe(IframeFragmentInfo),
     Image(Box<ImageFragmentInfo>),
     Canvas(Box<CanvasFragmentInfo>),
+    SVG(Box<SVGFragmentInfo>),
```
Comment (pcwalton) [inline components/layout/fragment.rs]: nit: name this `Svg`; we follow Rust conventions in layout, not DOM ones

File: components/layout/fragment.rs [hunk 2]
```diff
@@ -356,6 +360,44 @@ impl CanvasFragmentInfo {
...
+                                        ctx: &SharedStyleContext)
+                                        -> SVGFragmentInfo {
+        SVGFragmentInfo {
+            replaced_image_fragment_info: ReplacedImageFragmentInfo::new(node, ctx),
+            dom_width: Au::from_px(data.width as i32),
+            dom_height: Au::from_px(data.height as i32),
+        }
+    }
+
+    /// Returns the original inline-size of the canvas.
```
Comment (pcwalton) [inline components/layout/fragment.rs]: Presumably you mean "SVG element" instead of "canvas"?

**Comment by splav** [comment]
Note: this is WIP. As it required a lot of small modifications in different places I definitely missed something.

**Comment by splav** [comment]
Yep... I should stop changing code before 1.30am :)

5 окт. 2016 г. 5:17 AM пользователь "Ravi Shankar" <notifications@github.com>
написал:

File: components/script/dom/create.rs [hunk 3]
```diff
@@ -76,21 +76,44 @@ use dom::htmltrackelement::HTMLTrackElement;
...
+    macro_rules! make(
+        ($ctor:ident) => ({
+            let obj = $ctor::new(name.local, prefix, document);
+            Root::upcast(obj)
+        });
+        ($ctor:ident, $($arg:expr),+) => ({
+            let obj = $ctor::new(name.local, prefix, document, $($arg),+);
+            Root::upcast(obj)
+        })
+    );
```
Comment (splav) [inline components/script/dom/create.rs]: Actually it's not so obvious. If I move macro out of scope I'll lose name.local, prefix, document local vars and should make them arguments. That makes macro mostly useless.

**Comment by jdm** [comment]
That's just #13593. This shouldn't have any effect on other tests because all the code is preffed off by default.

File: components/script/dom/svggraphicselement.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    fn super_type(&self) -> Option<&VirtualMethods> {
+        Some(self.upcast::<SVGElement>() as &VirtualMethods)
+    }
+
+    fn attribute_mutated(&self, attr: &Attr, mutation: AttributeMutation) {
+        self.super_type().unwrap().attribute_mutated(attr, mutation);
+    }
+
+    fn bind_to_tree(&self, tree_in_doc: bool) {
+        if let Some(ref s) = self.super_type() {
```
Comment (Ms2ger) [inline components/script/dom/svggraphicselement.rs]: Neither of those methods need to be overridden, I think.

File: components/script/dom/svgelement.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+impl VirtualMethods for SVGElement {
+    fn super_type(&self) -> Option<&VirtualMethods> {
+        Some(self.upcast::<Element>() as &VirtualMethods)
+    }
+
+    fn attribute_mutated(&self, attr: &Attr, mutation: AttributeMutation) {
+        self.super_type().unwrap().attribute_mutated(attr, mutation);
+    }
+
+    fn bind_to_tree(&self, tree_in_doc: bool) {
```
Comment (Ms2ger) [inline components/script/dom/svgelement.rs]: Drop those two overrides.

File: components/script/dom/svgelement.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+use dom::bindings::inheritance::Castable;
+use dom::bindings::js::Root;
+use dom::bindings::str::DOMString;
+use dom::document::Document;
+use dom::element::AttributeMutation;
+use dom::element::Element;
+use dom::node::Node;
+use dom::virtualmethods::VirtualMethods;
+use string_cache::Atom;
+use style::element_state::*;
```
Comment (Ms2ger) [inline components/script/dom/svgelement.rs]: Please avoid glob imports

File: components/script/dom/create.rs (hunk 3 above)
Comment (Ms2ger) [inline components/script/dom/create.rs]: Alright, let's keep it duplicated for now.

**Review by Ms2ger — CHANGES_REQUESTED** [review]
DOM parts seem generally OK to me.

**Review by Ms2ger — APPROVED** [review]

**Comment by Ms2ger** [comment]
r+ on the DOM parts; it's not clear to me if everything else has been reviewed.

File: components/layout/fragment.rs [hunk 7]
```diff
@@ -356,6 +360,44 @@ impl CanvasFragmentInfo {
     }
 }
 
+#[derive(Clone)]
+pub struct SVGFragmentInfo {
```
Comment (pcwalton) [inline components/layout/fragment.rs]: nit: name this `SvgFragmentInfo`

**Review by pcwalton — APPROVED** [review]
r=me with nits

**Comment by jdm** [comment]
@bors-servo: retry
* #13360

**Comment by jdm** [comment]
@bors-servo: retry
* #13645

**Pull request merged**